src/coreclr/vm/ceemain.cpp

```diff
@@ -898,6 +867,44 @@ void EEStartupHelper()
 
         IfFailGo(hr);
 
+#ifdef PROFILING_SUPPORTED
+        // Initialize the profiling services.
+        hr = ProfilingAPIUtility::InitializeProfiling();
+
+        _ASSERTE(SUCCEEDED(hr));
+        IfFailGo(hr);
+#endif // PROFILING_SUPPORTED
+
+        InitializeExceptionHandling();
+
+        //
+        // Install our global exception filter
+        //
+        if (!InstallUnhandledExceptionFilter())
+        {
+            IfFailGo(E_FAIL);
+        }
+
+#ifdef TARGET_WINDOWS
+        // g_pGCHeap->Initialize() above could take nontrivial time, so by now the finalizer thread
```

Comment:
Can we end up creating the finalizer thread too soon on non-Windows?

For example, can the finalizer `Thread` be registered in thread store before `ProfilingAPIUtility::InitializeProfiling` and miss profiler notifications that are done as part of the registration?

Reply:
If `InitializeProfiling` does not need `Thread` object for the current thread, it can be moved earlier. 

The comments inside `InitializeProfiling` say that there are asserts somewhere that finalizer thread has started. So I wonder - are there tests for profiler attach?

Reply:
> The comments inside InitializeProfiling say that there are asserts somewhere that finalizer thread has started.

These comments refer to the old profiler attach implementation. I am deleting them in https://github.com/dotnet/runtime/pull/115680

> So I wonder - are there tests for profiler attach?

Yes, look for "Attach" under https://github.com/dotnet/runtime/tree/main/src/tests/profiler. (The tests won't hit every subtle race condition though.)

Reply:
I've moved the profiler initializtion earlier. If it needs to attach a Thread, then it would hit an assert in `OsAttachThread`. There is an assert there to fail deterministically.

I was mostly wondering if we test scenarios beyond "profiler is not involved at all".

Reply:
With default test set there are no failures, so based on that we could say profiler initializtion does not create/attach Thread objects.

If the profiler tests you've pointed too run in default test set, it may be good enough. If we need to run some other leg - like outer loop, perhaps makes sense to do that. Maybe makes sense to do that regardless.